frontend/main.js

```diff
@@ -321,34 +365,126 @@ async function startApp () {
       }
       sbp('okTurtles.events/off', CONTRACT_IS_SYNCING, initialSyncFn)
       sbp('okTurtles.events/on', CONTRACT_IS_SYNCING, syncFn.bind(this))
-      sbp('okTurtles.events/on', LOGIN, async () => {
-        this.ephemeral.finishedLogin = 'yes'
+      sbp('okTurtles.events/on', LOGOUT, () => {
+        // TODO: This is to be done by the SW
+        Promise.all([
+          sbp('chelonia/reset'),
+          sbp('gi.db/settings/delete', 'CHELONIA_STATE')
+        ]).catch(e => {
+          console.error('Logout event: error deleting Chelonia state')
+        })
+      })
+      sbp('okTurtles.events/on', LOGIN, async ({ identityContractID, encryptionParams, state }) => {
+        try {
+          const vuexState = sbp('state/vuex/state')
+          if (vuexState.loggedIn && vuexState.loggedIn.identityContractID !== identityContractID) {
+            throw new Error('Received login event but there already is an active session')
+          }
+          const cheloniaState = cloneDeep(await sbp('chelonia/rootState'))
+          if (state) {
+          // TODO Do this in a cleaner way
+          // Exclude contracts from the state
+            Object.keys(state).forEach(k => {
+              if (k.startsWith('z9br')) {
+                delete state[k]
+              }
+            })
+            Object.keys(cheloniaState.contracts).forEach(k => {
+              if (cheloniaState[k]) {
+                state[k] = cheloniaState[k]
```

Comment:
`Vue.set`?

Reply:
It's not needed because it's used to replace the state.